vaporetto/src/sentence.rs

```diff
@@ -82,182 +81,223 @@ pub enum BoundaryType {
+    pub(crate) score_padding: usize,
+    pub(crate) char_pma_states: Vec<u32>,
+    pub(crate) type_pma_states: Vec<u32>,
+    pub(crate) tags: Vec<Option<Cow<'b, str>>>,
+    pub(crate) n_tags: usize,
+    predictor: Option<&'b Predictor>,
+    str_to_char_pos: Vec<usize>,
+    char_to_str_pos: Vec<usize>,
 }
 
-impl TagScores {
-    /// Clears scores.
-    pub fn clear(&mut self) {
-        self.left_scores.clear();
-        self.right_scores.clear();
-        self.self_scores.clear();
-    }
-
-    /// Initializes score arrays.
+impl<'a, 'b> Default for Sentence<'a, 'b> {
+    /// Creates a new [`Sentence`] from a space.
     ///
-    /// # Arguments
+    /// # Examples
     ///
-    /// * `n_chars` - Length of characters in code points.
-    /// * `n_tags` - The number of tags.
-    #[allow(dead_code)]
-    pub fn init(&mut self, n_chars: usize, n_tags: usize) {
-        self.clear();
-        self.left_scores.resize(n_chars * n_tags, 0);
-        self.right_scores.resize(n_chars * n_tags, 0);
-        self.self_scores.resize(n_chars, None);
-    }
-}
-
-/// Sentence with boundary annotations.
-#[derive(Debug, PartialEq, Clone)]
-pub struct Sentence {
-    pub(crate) text: String,
-    pub(crate) chars: Vec<char>,
-    pub(crate) str_to_char_pos: Vec<usize>,
-    pub(crate) char_to_str_pos: Vec<usize>,
-    pub(crate) char_type: Vec<u8>,
-    pub(crate) boundaries: Vec<BoundaryType>,
-    pub(crate) boundary_scores: Vec<i32>,
-    pub(crate) tag_scores: TagScores,
-    pub(crate) tags: Vec<Option<Arc<String>>>,
-}
-
-impl Sentence {
-    fn internal_new(
-        text: String,
-        chars: Vec<char>,
-        boundaries: Vec<BoundaryType>,
-        tags: Vec<Option<Arc<String>>>,
-    ) -> Self {
-        let mut s = Self {
-            text,
-            chars,
-            str_to_char_pos: vec![],
-            char_to_str_pos: vec![],
-            char_type: vec![],
-            boundaries,
+    /// ```
+    /// use vaporetto::Sentence;
+    ///
+    /// let s = Sentence::default();
+    ///
+    /// assert_eq!(" ", s.as_raw_text());
+    /// assert_eq!(0, s.n_tags());
+    /// ```
+    fn default() -> Self {
+        Self {
+            text: Cow::Borrowed(" "),
+            char_types: vec![CharacterType::Other as u8],
+            boundaries: vec![],
             boundary_scores: vec![],
-            tag_scores: TagScores::default(),
-            tags,
-        };
-        s.update_common_info();
-        s
+            score_padding: 0,
+            char_pma_states: vec![],
+            type_pma_states: vec![],
+            tags: vec![],
+            n_tags: 0,
+            predictor: None,
+            str_to_char_pos: vec![0, 1],
+            char_to_str_pos: vec![0, 1],
+        }
     }
+}
 
-    fn clear(&mut self) {
-        self.text.clear();
-        self.text.push(' ');
-        self.chars.clear();
-        self.chars.push(' ');
+impl<'a, 'b> Sentence<'a, 'b> {
+    fn set_default(&mut self) {
```

Comment:
`Sentence` does not implement `PartialEq` because the comparison process is rather heavy, so I changed to call `set_default()` in `Default::default()`.